zim/gui/pageview/__init__.py

```diff
@@ -494,63 +521,93 @@ def image_file_load_pixels(file, width_override=-1, height_override=-1):
-
-		pixbuf = rotate_pixbuf(pixbuf)
-
-	except:
-		logger.debug('GTK failed to read image, using Pillow fallback: %s', file.path)
-
-		from PIL import Image # load Pillow only if necessary
-
-		with Image.open(file.path) as img_pil:
-
-			# resize if a specific size was requested
-			if b_size_override:
-				logger.debug('PIL resizing %s %s', width_override, height_override)
-				img_pil = img_pil.resize((width_override, height_override))
+	if need_switch_to_fallback:
+		try:
+			# load Pillow only if necessary
+			# noinspection PyUnresolvedReferences
+			from PIL import Image, UnidentifiedImageError
+		except ImportError:
+			pass
+		else:
+			try:
+				with Image.open(file.path) as image:
+					pixbuf = _convert_pillow_image_to_pixbuf(image)
+			except UnidentifiedImageError:
+				logger.debug(f'Pillow failed to read image: {file.path}')
+			else:
+				need_switch_to_fallback = False
+
+	if need_switch_to_fallback:
+		error_message = f'No available fallback for load this image: {file.path}'
+		logger.debug(error_message)
+		raise TypeError(error_message)
+
+	# Let's try to find and remember the orientation before scaling,
+	# 	because we lose metadata when changing images.
+	orientation = None
+	mimetype = file.mimetype()
+	if mimetype in {'image/jpeg', 'image/tiff'}:
+		# Gtk can detect orientation in jpeg|tiff images only
+		# See docs: https://docs.gtk.org/gdk-pixbuf/method.Pixbuf.get_option.html#description
+		orientation = pixbuf.get_option('orientation')
+	if mimetype in {'image/webp', 'image/png'}:
+		# if possible, we will find orientation of the image using Pillow,
+		# 	if it is not available, we will display image it as is.
+		try:
+			# noinspection PyUnresolvedReferences
+			from PIL import Image, UnidentifiedImageError, __version__ as pillow_version_string
+		except ImportError:
+			pass
+		else:
+			pillow_version = tuple(map(int, pillow_version_string.split('.')))
+			try:
+				with Image.open(file.path) as image:
+					if pillow_version >= (6, 0, 0):
+						# https://pillow.readthedocs.io/en/stable/releasenotes/6.0.0.html#added-exif-class
+						exif = image.getexif()
+					else:
+						# noinspection PyUnresolvedReferences,PyProtectedMember
+						exif = image._getexif()  # noqa: WPS437
+						if not exif:
+							exif = {}  # noqa: WPS437
+			except UnidentifiedImageError:
+				logger.debug(f'Pillow failed to read image: {file.path}')
+			else:
+				orientation_tag_id = 274
+				orientation = exif.get(orientation_tag_id)
+	orientation = int(orientation) if orientation else 1
 
-			# check if there is an alpha channel
-			if img_pil.mode == 'RGB':
-				has_alpha = False
-			elif img_pil.mode == 'RGBA':
-				has_alpha = True
+	w, h = pixbuf.get_width(), pixbuf.get_height()
+	b_size_override = width_override > 0 or height_override > 0
+	if b_size_override and (width_override <= 0 or height_override <= 0):
+		if orientation in {5, 6, 7, 8}:
```

Comment:
Before, we resized prior to rotation. I do agree that taking the orientation into account makes sense, but does this break the dimensions of all existing exif-rotated images?

Reply:
Nope.
We need to swap the height with the width to get the correct image proportions for orientation in {5, 6, 7, 8}
I wrote a small test case to check if it works https://github.com/Toggoren/PainNotes/tree/master/PainNotes

Reply:
Sorry, I meant whether resized images appear the same size before and after applying this patch - though now I realize that might be impossible to test given that rotation was broken before :D